File: riveredge-frontend/src/apps/kuaizhizao/services/demand.ts
/**
 * 统一需求管理API服务
 *
 * 提供统一需求管理相关的API调用接口，支持销售预测和销售订单两种需求类型。
 *
 * @author Luigi Lu
 * @date 2025-01-14
 */

import { apiRequest } from '../../../services/api';

/**
 * 统一需求接口定义
 */
export enum DemandStatus {
  DRAFT = 'DRAFT',
  PENDING_REVIEW = 'PENDING_REVIEW',
  AUDITED = 'AUDITED',
  REJECTED = 'REJECTED',
  CONFIRMED = 'CONFIRMED'
}

export enum ReviewStatus {
  PENDING = 'PENDING',
  APPROVED = 'APPROVED',
  REJECTED = 'REJECTED'
}

export interface Demand {
  id?: number;
  uuid?: string;
  tenant_id?: number;
  demand_code?: string;
  demand_type?: 'sales_forecast' | 'sales_order';
  demand_name?: string;
  business_mode?: 'MTS' | 'MTO';
  start_date?: string;
  end_date?: string;
  customer_id?: number;
  customer_name?: string;
  customer_contact?: string;
  customer_phone?: string;
  forecast_period?: string;
  order_date?: string;
  delivery_date?: string;
  total_quantity?: number;
  total_amount?: number;
  status?: string;
  reviewer_id?: number;
  reviewer_name?: string;
  review_time?: string;
  review_status?: string;
  review_remarks?: string;
  submit_time?: string;
  salesman_id?: number;
  salesman_name?: string;
  shipping_address?: string;
  shipping_method?: string;
  payment_terms?: string;
  source_id?: number;
  source_type?: string;
  source_code?: string;
  priority?: number;
  pushed_to_computation?: boolean;
  computation_id?: number;
  computation_code?: string;
  notes?: string;
  is_active?: boolean;
  created_by?: number;
  updated_by?: number;
  created_at?: string;
  updated_at?: string;
  items?: DemandItem[];
  /** 生命周期（后端计算，列表/详情接口返回，供 UniLifecycleStepper 展示） */
  lifecycle?: {
    current_stage_key?: string;
    current_stage_name?: string;
    status?: 'success' | 'exception' | 'normal' | 'active';
    main_stages?: { key: string; label: string; status: 'done' | 'active' | 'pending' }[];
    sub_stages?: { key: string; label: string; status: 'done' | 'active' | 'pending' }[];
  };
  duration_info?: {
    created_at?: string;
    submit_time?: string;
    review_time?: string;
    duration_to_submit?: number | null;
    duration_to_review?: number | null;
    duration_submit_to_review?: number | null;
  };
}

/**
 * 统一需求明细接口定义
 */
export interface DemandItem {
  id?: number;
  uuid?: string;
  tenant_id?: number;
  demand_id?: number;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  material_spec?: string;
  material_unit?: string;
  required_quantity?: number;
  forecast_date?: string;
  forecast_month?: string;
  historical_sales?: number;
  historical_period?: string;
  confidence_level?: number;
  forecast_method?: string;
  delivery_date?: string;
  delivered_quantity?: number;
  remaining_quantity?: number;
  unit_price?: number;
  item_amount?: number;
  delivery_status?: string;
  work_order_id?: number;
  work_order_code?: string;
  notes?: string;
  created_at?: string;
  updated_at?: string;
}

/**
 * 需求列表查询参数
 */
export interface DemandListParams {
  skip?: number;
  limit?: number;
  demand_type?: 'sales_forecast' | 'sales_order';
  status?: string;
  pushed_to_computation?: boolean;
  business_mode?: 'MTS' | 'MTO';
  review_status?: string;
  start_date?: string;
  end_date?: string;
}

/**
 * 需求列表响应
 */
export interface DemandListResponse {
  data: Demand[];
  total: number;
  success: boolean;
}

/**
 * 获取需求列表
 */
export async function listDemands(params: DemandListParams = {}): Promise<DemandListResponse> {
  return apiRequest<DemandListResponse>('/apps/kuaizhizao/demands', {
    method: 'GET',
    params,
  });
}

/**
 * 获取需求详情
 */
export async function getDemand(id: number, includeItems: boolean = false, includeDuration: boolean = false): Promise<Demand & { duration_info?: any }> {
  return apiRequest<Demand & { duration_info?: any }>(`/apps/kuaizhizao/demands/${id}`, {
    method: 'GET',
    params: { include_items: includeItems, include_duration: includeDuration },
  });
}

/**
 * 创建需求
 */
export async function createDemand(data: Partial<Demand>): Promise<Demand> {
  return apiRequest<Demand>('/apps/kuaizhizao/demands', {
    method: 'POST',
    data,
  });
}

/**
 * 更新需求
 */
export async function updateDemand(id: number, data: Partial<Demand>): Promise<Demand> {
  return apiRequest<Demand>(`/apps/kuaizhizao/demands/${id}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 提交需求
 */
export async function submitDemand(id: number): Promise<Demand> {
  return apiRequest<Demand>(`/apps/kuaizhizao/demands/${id}/submit`, {
    method: 'POST',
  });
}

/**
 * 审核通过需求
 */
export async function approveDemand(id: number): Promise<Demand> {
  return apiRequest<Demand>(`/apps/kuaizhizao/demands/${id}/approve`, {
    method: 'POST',
  });
}

/**
 * 驳回需求
 */
export async function rejectDemand(id: number, rejectionReason: string): Promise<Demand> {
  return apiRequest<Demand>(`/apps/kuaizhizao/demands/${id}/reject`, {
    method: 'POST',
    params: { rejection_reason: rejectionReason },
  });
}

/**
 * 添加需求明细
 */
export async function addDemandItem(demandId: number, data: Partial<DemandItem>): Promise<DemandItem> {
  return apiRequest<DemandItem>(`/apps/kuaizhizao/demands/${demandId}/items`, {
    method: 'POST',
    data,
  });
}

/**
 * 更新需求明细
 */
export async function updateDemandItem(demandId: number, itemId: number, data: Partial<DemandItem>): Promise<DemandItem> {
  return apiRequest<DemandItem>(`/apps/kuaizhizao/demands/${demandId}/items/${itemId}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 删除需求明细
 */
export async function deleteDemandItem(demandId: number, itemId: number): Promise<void> {
  return apiRequest<void>(`/apps/kuaizhizao/demands/${demandId}/items/${itemId}`, {
    method: 'DELETE',
  });
}

/**
 * 批量创建需求响应
 */
export interface BatchCreateDemandsResponse {
  success: boolean;
  success_count: number;
  failure_count: number;
  errors: Array<{
    row: number;
    error: string;
    data: any;
  }>;
  created_demands: Demand[];
}

/**
 * 批量创建需求
 */
export async function batchCreateDemands(demands: Partial<Demand>[]): Promise<BatchCreateDemandsResponse> {
  return apiRequest<BatchCreateDemandsResponse>('/apps/kuaizhizao/demands/batch', {
    method: 'POST',
    data: demands,
  });
}

/**
 * 下推需求到物料需求运算响应
 */
export interface PushToComputationResponse {
  success: boolean;
  message: string;
  demand_code: string;
  computation_code: string;
  note?: string;
}

/**
 * 下推需求到物料需求运算
 */
export async function pushDemandToComputation(demandId: number): Promise<PushToComputationResponse> {
  return apiRequest<PushToComputationResponse>(`/apps/kuaizhizao/demands/${demandId}/push-to-computation`, {
    method: 'POST',
  });
}

/** 需求重算历史项（与后端 list_demand_recalc_history 返回一致） */
export interface DemandRecalcHistoryItem {
  id: number;
  recalc_at?: string;
  trigger_type?: string;
  source_type?: string;
  source_id?: number;
  trigger_reason?: string;
  snapshot_id?: number;
  operator_id?: number;
  result?: string;
  message?: string;
}

/** 需求快照项（与后端 list_demand_snapshots 返回一致） */
export interface DemandSnapshotItem {
  id: number;
  snapshot_type?: string;
  snapshot_at?: string;
  trigger_reason?: string;
  demand_snapshot?: Record<string, any>;
  demand_items_snapshot?: any[];
}

/**
 * 获取需求重算历史列表
 */
export async function listDemandRecalcHistory(
  demandId: number,
  params?: { limit?: number }
): Promise<DemandRecalcHistoryItem[]> {
  return apiRequest<DemandRecalcHistoryItem[]>(`/apps/kuaizhizao/demands/${demandId}/recalc-history`, {
    method: 'GET',
    params,
  });
}

/**
 * 获取需求快照列表
 */
export async function listDemandSnapshots(
  demandId: number,
  params?: { limit?: number }
): Promise<DemandSnapshotItem[]> {
  return apiRequest<DemandSnapshotItem[]>(`/apps/kuaizhizao/demands/${demandId}/snapshots`, {
    method: 'GET',
    params,
  });
}

/** 清理孤儿需求接口返回 */
export interface CleanOrphanDemandsResponse {
  cleaned_count: number;
  demand_ids: number[];
}

/**
 * 清理孤儿需求（直接物理删除：来源单据已删除的需求）
 */
export async function cleanOrphanDemands(): Promise<CleanOrphanDemandsResponse> {
  return apiRequest<CleanOrphanDemandsResponse>('/apps/kuaizhizao/demands/clean-orphans', {
    method: 'POST',
  });
}

File: riveredge-backend/src/apps/kuaizhizao/models/demand.py
"""
统一需求模型

提供统一的需求数据模型定义，支持销售预测和销售订单两种需求类型。

根据《☆ 用户使用全场景推演.md》的设计理念，将销售预测和销售订单统一为需求管理，
作为整个业务流程的源头。

Author: Luigi Lu
Date: 2025-01-14
"""

from tortoise import fields
from core.models.base import BaseModel


class Demand(BaseModel):
    """
    统一需求模型

    用于统一管理销售预测（MTS模式）和销售订单（MTO模式）两种需求类型。
    作为整个业务流程的源头，支持需求下推到物料需求运算。

    需求类型（demand_type）：
    - sales_forecast: 销售预测（MTS模式）
    - sales_order: 销售订单（MTO模式）
    
    注意：继承自BaseModel，自动获得id、uuid、tenant_id、created_at、updated_at字段
    """
    demand_code = fields.CharField(max_length=50, description="需求编码")
    
    # 需求类型（sales_forecast 或 sales_order）
    demand_type = fields.CharField(max_length=20, description="需求类型")
    
    # 需求基本信息（通用字段）
    demand_name = fields.CharField(max_length=200, description="需求名称")
    
    # 业务模式（MTS 或 MTO，根据需求类型自动识别）
    business_mode = fields.CharField(max_length=20, description="业务模式（MTS/MTO）")
    
    # 时间范围（通用）
    start_date = fields.DateField(description="开始日期")
    end_date = fields.DateField(null=True, description="结束日期（销售订单可为空）")
    
    # 客户信息（销售订单专用，销售预测可为空）
    customer_id = fields.IntField(null=True, description="客户ID")
    customer_name = fields.CharField(max_length=200, null=True, description="客户名称")
    customer_contact = fields.CharField(max_length=100, null=True, description="客户联系人")
    customer_phone = fields.CharField(max_length=20, null=True, description="客户电话")
    
    # 销售预测专用字段
    forecast_period = fields.CharField(max_length=20, null=True, description="预测周期（销售预测专用）")
    
    # 销售订单专用字段
    order_date = fields.DateField(null=True, description="订单日期（销售订单专用）")
    delivery_date = fields.DateField(null=True, description="交货日期（销售订单专用）")
    
    # 金额信息（通用）
    total_quantity = fields.DecimalField(max_digits=10, decimal_places=2, default=0, description="总数量")
    total_amount = fields.DecimalField(max_digits=12, decimal_places=2, default=0, description="总金额")
    
    # 状态（通用）
    status = fields.CharField(max_length=20, default="草稿", description="需求状态")
    
    # 时间节点记录（用于耗时统计）
    submit_time = fields.DatetimeField(null=True, description="提交时间")
    
    # 审核信息（通用）
    reviewer_id = fields.IntField(null=True, description="审核人ID")
    reviewer_name = fields.CharField(max_length=100, null=True, description="审核人姓名")
    review_time = fields.DatetimeField(null=True, description="审核时间")
    review_status = fields.CharField(max_length=20, default="待审核", description="审核状态")
    review_remarks = fields.TextField(null=True, description="审核备注")
    
    # 销售信息（销售订单专用）
    salesman_id = fields.IntField(null=True, description="销售员ID")
    salesman_name = fields.CharField(max_length=100, null=True, description="销售员姓名")
    
    # 物流信息（销售订单专用）
    shipping_address = fields.TextField(null=True, description="收货地址")
    shipping_method = fields.CharField(max_length=50, null=True, description="发货方式")
    payment_terms = fields.CharField(max_length=100, null=True, description="付款条件")
    
    # 需求关联信息
    source_id = fields.IntField(null=True, description="来源ID（用于关联原始销售预测或销售订单）")
    source_type = fields.CharField(max_length=50, null=True, description="来源类型")
    source_code = fields.CharField(max_length=50, null=True, description="来源编码")
    
    # 下推信息
    pushed_to_computation = fields.BooleanField(default=False, description="是否已下推到需求计算")
    computation_id = fields.IntField(null=True, description="关联的需求计算ID")
    computation_code = fields.CharField(max_length=50, null=True, description="关联的需求计算编码")
    
    notes = fields.TextField(null=True, description="备注")
    is_active = fields.BooleanField(default=True, description="是否有效")
    created_by = fields.IntField(null=True, description="创建人ID")
    updated_by = fields.IntField(null=True, description="更新人ID")
    deleted_at = fields.DatetimeField(null=True, description="删除时间")
    
    class Meta:
        table = "apps_kuaizhizao_demands"
        table_description = "快格轻制造 - 统一需求"
        indexes = [
            ("tenant_id", "demand_type"),
            ("tenant_id", "status"),
            ("tenant_id", "business_mode"),
            ("start_date",),
            ("end_date",),
            ("delivery_date",),
            ("customer_id",),
            ("demand_code",),
        ]
    
    class PydanticMeta:
        exclude = ["deleted_at"]
